plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,317 @@
+   that in this situation:
+   ```
+   do nms <- genFreshNames ss
+      nms' <- genFreshNames ss
+   ```
+   the names in `nms` and `nms'` don't overlap.
+
+   Instead, what you are supposed to do is locally use the names in `nms` and `nms'` to
+   define generators that use them. This is done with functions like `bindTyName` and `bindTmName`:
+   ```
+   genLam ma mb = do
+      x <- genFreshName "x"
+      sizeSplit 1 7 (maybe (genType Star) return ma)
+                    --      v--- LOOK HERE!
+                    (\ a -> bindTmName x a . withNoEscape $ genTerm mb) $ \ a (b, body) ->
+                    --      ^--- LOOK HERE!
+                    TyFun () a b, LamAbs () x a body)
+   ```
+-}
+
+-- | Generate a fresh name. See Note [Warning about generating fresh names].
+genFreshName :: String -> GenTm Name
+genFreshName s = head <$> genFreshNames [s]
+
+-- | Generate one fresh name per string in the input list.
+-- names don't overlap. See Note [Warning about generating fresh names].
+genFreshNames :: [String] -> GenTm [Name]
+genFreshNames ss = do
+  used <- getUniques
+  let i = fromEnum $ Set.findMax $ Set.insert (Unique 0) used
+      js = [ j | j <- [1..i], not $ Unique j `Set.member` used ]
+      is = js ++ take (length ss + 10) [i+1..]
+  is' <- liftGen $ QC.shuffle is
+  return [Name (fromString $ s ++ show j) (toEnum j) | (s, j) <- zip ss is']
+
+-- | See `genFreshName`
+genFreshTyName :: String -> GenTm TyName
+genFreshTyName s = TyName <$> genFreshName s
+
+-- | See `genFreshNames`
+genFreshTyNames :: [String] -> GenTm [TyName]
+genFreshTyNames ss = map TyName <$> genFreshNames ss
+
+-- | Generate a name that likely overlaps with existing names on purpose. If there are no existing
+-- names, generate a fresh name. This function doesn't distinguish between the type- and term-level
+-- scopes, hence it may generate a 'Name' \"clashing\" with a previously generated 'TyName' and not
+-- clashing with any previously generated 'Name'.
+genLikelyNotFreshName :: String -> GenTm Name
+genLikelyNotFreshName s = do
+  used <- Set.toList <$> getUniques
+  case used of
+    [] -> genFreshName s
+    _  -> liftGen $ elements [ Name (fromString $ s ++ show (unUnique i)) i | i <- used ]
+
+-- | Generate a fresh name most (a bit more than 75%) of the time and otherwise
+-- generate an already bound name. When there are no bound names generate a fresh name.
+genMaybeFreshName :: String -> GenTm Name
+genMaybeFreshName s = frequency [(3, genFreshName s), (1, genLikelyNotFreshName s)]
+
+-- | See `genMaybeFreshName`
+genMaybeFreshTyName :: String -> GenTm TyName
+genMaybeFreshTyName s = TyName <$> genMaybeFreshName s
+
+-- | Bind a type name to a kind and avoid capturing free type variables.
+bindTyName :: TyName -> Kind () -> GenTm a -> GenTm a
+bindTyName x k = local $ \ e -> e
+    { geTypes = Map.insert x k (geTypes e)
+    , geTerms = Map.filter (\ty -> not $ x `Set.member` setOf ftvTy ty) (geTerms e)
```

Comment:
> you're removing things from the environment that match the new varaible?

that reference the now shadowed variable in their type.

Reply:
Okay, that makes a sort of sense but it seems potentially buggy:
- `geTerms` is _not_ the set of terms that are in scope, but the set of terms in scope that can be correctly inserted into the program. Similarly for the others. This should definitely be documented!
- Fresh name generation relies on the invariant that all the generated uniques appear in one of the in-scope maps, but this invariant is violated by what's happening here, since when you add a fresh type variable you may lose an arbitrary number of term variables and also some constructor variables for any datatypes that get removed.

Or maybe this is fine? If you re-generate one of the removed names you'll just shadow it, and maybe that's fine. But then the use of "fresh name" seems confusing since it means something different to what it means elsewhere.

Reply:
Those are very good points... So `genFreshName` is in fact `getLikelyFreshName`. Definitely worth documenting, I'll do it.

@MaximilianAlgehed so it looks like `genFreshName` can generate a non-fresh name. How do you feel about it, can that break some invariant or something or is it fine?

Reply:
Should be fine, but maybe worth looking in to anyway.

Reply:
I don't mean that it will have bad consequences, but that it might be morally better to have a function that is guaranteed to generate a fresh name.